README.md
# ChatGpt Proxy
An unofficial Golang program that provides both ChatGPT access token and API proxy simultaneously.

一个非官方的Golang程序，同时提供ChatGPT访问令牌和API代理。

## 配置文件
```yaml
host: ""
port: 55555
# 代理 	http://username:password@host:port 或 socks5://username:password@host:port
proxy-url: "http://127.0.0.1:7890"
openai-proxy:
  # 是否开启
  enable: true
  # 账号密码 可选
  email: ""
  password: ""
  # plus账号cookie 解除限制 可选
  puid: ""
  # api前缀
  prefix: "openai-proxy"
api-proxy:
  # 是否开启
  enable: true
  # api前缀
  prefix: "api-proxy"
```
## 快速入门
[根据自己系统版本下载release](https://github.com/popcell/chatgpt-proxy/releases)
填写config配置文件
```shell
# login to get access token
./chatgpt-proxy -c login 
# run proxy server
./chatgpt-proxy

# openai_proxy http://localhost:55555/openai-proxy
# api_proxy http://localhost:55555/api-proxy
```
## docker部署
```shell
wget https://raw.githubusercontent.com/popcell/chatgpt-proxy/master/config.yaml
# 修改config.yaml
docker build -t chatgpt-proxy .
docker run -d --name chatgpt-proxy --restart always -p 55555:55555 -v $(pwd)/config.yaml:/app/config.yaml chatgpt-proxy
```
## docker-compose部署
```shell
wget https://raw.githubusercontent.com/popcell/chatgpt-proxy/master/config.yaml
# 修改config.yaml
docker build -t chatgpt-proxy .
docker-compose up -d
```
## 参考
[通过tls指纹绕过cloudflare](https://github.com/acheong08/ChatGPT-Proxy-V4)

[IOS获取access_token](https://zhile.io/2023/05/19/how-to-get-chatgpt-access-token-via-pkce.html)

RELEASE_NOTES.md
# 第一个发布版本

utils/logger/logger.go
package logger

import (
	"fmt"
	"github.com/sirupsen/logrus"
	"path/filepath"
	"runtime"
)

var (
	log *logrus.Logger
)

func init() {
	log = logrus.New()
	log.SetLevel(logrus.InfoLevel)
	log.SetFormatter(&logrus.TextFormatter{
		ForceColors:     true,
		FullTimestamp:   true,
		TimestampFormat: "2006-01-02 15:04:05",
	})
}

func getFileFromPath(path string) string {
	return filepath.Base(filepath.Dir(path)) + "/" + filepath.Base(path)
}
func Ansi(colorString string) func(...interface{}) string {
	return func(args ...interface{}) string {
		return fmt.Sprintf(colorString, fmt.Sprint(args...))
	}
}

var (
	Green  = Ansi("\033[1;32m%s\033[0m")
	Yellow = Ansi("\033[1;33m%s\033[0m")
	Red    = Ansi("\033[1;31m%s\033[0m")
)

func Info(args ...interface{}) {
	_, file, line, _ := runtime.Caller(1)
	// 记录日志消息
	log.WithFields(logrus.Fields{
		"file": fmt.Sprintf("%s:%d", getFileFromPath(file), line),
	}).Info(Green(args...))
}

func Warn(args ...interface{}) {
	_, file, line, _ := runtime.Caller(1)
	// 记录日志消息
	log.WithFields(logrus.Fields{
		"file": fmt.Sprintf("%s:%d", getFileFromPath(file), line),
	}).Warn(Yellow(args...))
}

func Error(args ...interface{}) {
	_, file, line, _ := runtime.Caller(1)
	// 记录日志消息
	log.WithFields(logrus.Fields{
		"file": fmt.Sprintf("%s:%d", getFileFromPath(file), line),
	}).Error(Red(args...))
}
func Panic(args ...interface{}) {
	_, file, line, _ := runtime.Caller(1)
	// 记录日志消息
	log.WithFields(logrus.Fields{
		"file": fmt.Sprintf("%s:%d", getFileFromPath(file), line),
	}).Panic(Red(args...))
}


api/get_access_token.go
package api

import (
	"crypto/rand"
	"crypto/sha256"
	"crypto/tls"
	"encoding/base64"
	"encoding/json"
	"fmt"
	log "github.com/popcell/chatgpt-proxy/utils/logger"
	"golang.org/x/net/context"
	"golang.org/x/net/proxy"
	"net"
	"net/http"
	"net/http/cookiejar"
	"net/url"
	"os"
	"regexp"
	"strings"
	"time"
)

type Auth0 struct {
	SessionToken string
	Email        string
	Password     string
	ApiPrefix    string
}

var (
	client         *http.Client
	proxyTransport *http.Transport
)

func GenerateCodeVerifier() string {
	token := make([]byte, 32)
	_, err := rand.Read(token)
	if err != nil {
		panic(err)
	}
	encodedToken := base64.RawURLEncoding.EncodeToString(token)
	codeVerifier := strings.TrimRight(encodedToken, "=")
	return codeVerifier
}

func GenerateCodeChallenge(codeVerifier string) string {
	hash := sha256.New()
	hash.Write([]byte(codeVerifier))
	codeChallenge := base64.RawURLEncoding.EncodeToString(hash.Sum(nil))
	codeChallenge = strings.TrimRight(codeChallenge, "=")
	return codeChallenge
}
func CheckEmail(email string) bool {
	emailRegex := `^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,7}$`
	return regexp.MustCompile(emailRegex).MatchString(email)
}

func NewAuth0(email, password string) *Auth0 {
	if CheckEmail(email) == false {
		panic("Invalid email")
	}
	return &Auth0{
		Email:    email,
		Password: password,
	}
}
func (auth Auth0) run() {
	auth.GetLoginUrl()
}
func (auth Auth0) GetLoginUrl() {
	codeVerifier := GenerateCodeVerifier()
	codeChallenge := GenerateCodeChallenge(codeVerifier)
	reqUrl := "https://auth0.openai.com/authorize?client_id=pdlLIX2Y72MIl2rhLhTE9VV9bN905kBh&audience=https%3A%2F%2Fapi.openai.com%2Fv1&redirect_uri=com.openai.chat%3A%2F%2Fauth0.openai.com%2Fios%2Fcom.openai.chat%2Fcallback&scope=openid%20email%20profile%20offline_access%20model.request%20model.read%20organization.read%20offline&response_type=code&code_challenge=" + codeChallenge + "&code_challenge_method=S256&prompt=login"
	auth.Login(codeVerifier, reqUrl)
}

func (auth Auth0) Login(codeVerifier string, reqUrl string) {
	// 创建 HTTP 请求
	req, err := http.NewRequest(http.MethodGet, reqUrl, nil)
	// 设置请求头
	req.Header.Set("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36")
	req.Header.Set("Referer", "https://ios.chat.openai.com/")

	if err != nil {
		fmt.Println("Login req Error:", err)
		return
	}

	// 发送 HTTP 请求
	resp, err := client.Do(req)

	if err != nil {
		fmt.Println("Login Do Error:", err)
		return
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusOK {
		parsedURL, _ := url.Parse(resp.Request.URL.String())
		queryValues := parsedURL.Query()
		state := queryValues.Get("state")
		client.CheckRedirect = func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		}
		auth.CheckUsername(codeVerifier, state)
	} else {
		fmt.Println("Login Do Error")
	}

}

func (auth Auth0) CheckUsername(codeVerifier string, state string) {
	reqUrl := fmt.Sprintf("https://auth0.openai.com/u/login/identifier?state=%s", state)
	// 创建 HTTP 请求
	formData := url.Values{
		"state":                       {state},
		"username":                    {auth.Email},
		"js-available":                {"true"},
		"webauthn-available":          {"true"},
		"is-brave":                    {"false"},
		"webauthn-platform-available": {"false"},
		"action":                      {"default"},
	}

	req, err := http.NewRequest("POST", reqUrl, strings.NewReader(formData.Encode()))
	// 设置请求头
	req.Header.Set("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36")
	req.Header.Set("Referer", reqUrl)
	req.Header.Set("Origin", "https://auth0.openai.com")
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	if err != nil {
		fmt.Println("CheckUsername Error:", err)
		return
	}

	resp, err := client.Do(req)

	if err != nil {
		fmt.Println("CheckUsername Error:", err)
		return
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusFound {
		auth.CheckPassword(codeVerifier, state)
	} else {
		fmt.Println("CheckUsername Error")
	}

}
func (auth Auth0) CheckPassword(codeVerifier string, state string) {
	reqUrl := fmt.Sprintf("https://auth0.openai.com/u/login/password?state=%s", state)
	// 创建 HTTP 请求
	formData := url.Values{
		"state":    {state},
		"username": {auth.Email},
		"password": {auth.Password},
		"action":   {"default"},
	}

	req, err := http.NewRequest(http.MethodPost, reqUrl, strings.NewReader(formData.Encode()))
	if err != nil {
		fmt.Println("CheckPassword req Error:", err)
		return
	}

	// 设置请求头
	req.Header.Set("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36")
	req.Header.Set("Referer", reqUrl)
	req.Header.Set("Origin", "https://auth0.openai.com")
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	resp, err := client.Do(req)

	if err != nil {
		fmt.Println("CheckPassword Error:", err)
		return
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusFound {
		location := resp.Header.Get("Location")
		if !strings.HasPrefix(location, "/authorize/resume?") {
			fmt.Println("Login failed.", err)
			return
		}
		auth.GetLocation(codeVerifier, location, reqUrl)
	} else if resp.StatusCode == http.StatusBadRequest {
		fmt.Println("Wrong email or password.", err)
		return
	} else {
		fmt.Println("Error Login.", err)
		return
	}

}

func (auth Auth0) GetLocation(codeVerifier string, location string, ref string) {
	reqUrl := fmt.Sprintf("https://auth0.openai.com%s", location)

	req, err := http.NewRequest("GET", reqUrl, nil)
	if err != nil {
		fmt.Println("GetLocation Error:", err)
		return
	}
	// 设置请求头
	req.Header.Set("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36")
	req.Header.Set("Referer", ref)
	resp, err := client.Do(req)
	defer resp.Body.Close()
	if err != nil {
		fmt.Println("GetLocation Error:", err)
		return
	}
	if resp.StatusCode == 302 {
		location := resp.Header.Get("Location")
		if !strings.HasPrefix(location, "com.openai.chat://auth0.openai.com/ios/com.openai.chat/callback?") {
			fmt.Println("Login failed GetLocation", err)
			return
		}
		auth.GetAccessToken(codeVerifier, resp.Header.Get("Location"))
	} else if resp.StatusCode == 400 {
		fmt.Println("Wrong email or password.", err)
		return
	} else {
		fmt.Println("Error Login.", err)
		return
	}
}

func (auth Auth0) GetAccessToken(codeVerifier string, callbackUrl string) {

	parsedURL, _ := url.Parse(callbackUrl)
	queryValues := parsedURL.Query()
	if reqError, ok := queryValues["error"]; ok {
		errorStr := fmt.Sprintf("%s: %s", reqError[0], queryValues["error_description"][0])
		fmt.Println("errorStr", errorStr)
		return
	} else if _, ok := queryValues["code"]; !ok {
		fmt.Println("Error get code from callback url.")
		return
	}
	reqUrl := "https://auth0.openai.com/oauth/token"
	// 创建 HTTP 请求
	formData := url.Values{
		"redirect_uri":  {"com.openai.chat://auth0.openai.com/ios/com.openai.chat/callback"},
		"grant_type":    {"authorization_code"},
		"client_id":     {"pdlLIX2Y72MIl2rhLhTE9VV9bN905kBh"},
		"code":          {queryValues["code"][0]},
		"code_verifier": {codeVerifier},
	}

	req, err := http.NewRequest("POST", reqUrl, strings.NewReader(formData.Encode()))
	if err != nil {
		fmt.Println("GetAccessToken Error:", err)
		return
	}
	// 设置请求头
	req.Header.Set("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36")
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	resp, err := client.Do(req)

	if err != nil {
		fmt.Println("GetAccessToken Error:", err)
		return
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusOK {
		var result map[string]interface{}
		json.NewDecoder(resp.Body).Decode(&result)
		accessToken := result["access_token"].(string)
		expiresAt := time.Now().Add(time.Second * time.Duration(result["expires_in"].(float64))).Add(-5 * time.Minute)
		expiresIn := expiresAt.Format("2006-01-02T15:04:05.000Z")
		fmt.Println(accessToken)
		fmt.Println(expiresIn)
		err := os.WriteFile("access_token.txt", []byte(accessToken), 0644)
		if err != nil {
			fmt.Println("Error writing file:", err)
			return
		}
		fmt.Println("accessToken written access_token.txt successfully")
	} else {
		fmt.Println("Error check email.")
	}
	return
}
func InitConfig(proxyUrl string) {
	if strings.HasPrefix(proxyUrl, "http") {
		proxyTransport = &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
			Proxy: func(req *http.Request) (*url.URL, error) {
				return url.Parse(proxyUrl)
			},
		}
	} else if strings.HasPrefix(proxyUrl, "socks5") {
		u, err := url.Parse(proxyUrl)
		if err != nil {
			log.Error("InitApiProxyConfig error: ", err.Error())
			return
		}
		username := u.User.Username()
		password, _ := u.User.Password()
		host := u.Hostname()
		address := fmt.Sprintf("%s:%s", host, u.Port())
		dialer, err := proxy.SOCKS5("tcp", address, &proxy.Auth{User: username, Password: password}, proxy.Direct)
		proxyTransport = &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
			DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
				return dialer.Dial(network, addr)
			},
		}
		if err != nil {
			log.Error("InitApiProxyConfig error: ", err.Error())
		}
	}
	// 创建 cookie 容器
	jar, err := cookiejar.New(nil)
	if err != nil {
		fmt.Println("Cookiejar Error:", err)
		return
	}

	client = &http.Client{
		Transport: proxyTransport,
		Timeout:   time.Second * 100,
		Jar:       jar,
	}
}
func Login(email string, password string) {
	auth0 := NewAuth0(email, password)
	auth0.run()
}


api/openai_proxy/openai_proxy.go
package openai_proxy

import (
	http "github.com/bogdanfinn/fhttp"
	tlsClient "github.com/bogdanfinn/tls-client"
	"github.com/gin-gonic/gin"
	log "github.com/popcell/chatgpt-proxy/utils/logger"
	"io"
)

var (
	jar     = tlsClient.NewCookieJar()
	options = []tlsClient.HttpClientOption{
		tlsClient.WithTimeoutSeconds(360),
		tlsClient.WithClientProfile(tlsClient.Safari_IOS_16_0),
		tlsClient.WithNotFollowRedirects(),
		tlsClient.WithCookieJar(jar),
	}
	client, _ = tlsClient.NewHttpClient(tlsClient.NewNoopLogger(), options...)
	userAgent = "Mozilla/5.0 (iPhone; CPU iPhone OS 14_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.2 Mobile/15E148 Safari/604.1"
	Puid      = ""
)

func InitOpenAiConfig(proxy string, puid string) {
	// http://username:password@host:port 或 socks5://username:password@host:port
	err := client.SetProxy(proxy)
	if err != nil {
		log.Error("proxy set error: ", err.Error())
	}
	Puid = puid
}
func OpenAiProxy(c *gin.Context) {
	// Remove _cfuvid cookie from session
	jar.SetCookies(c.Request.URL, []*http.Cookie{})
	var url string
	var err error
	var requestMethod string
	var request *http.Request
	var response *http.Response

	if c.Request.URL.RawQuery != "" {
		url = "https://chat.openai.com/backend-api" + c.Param("path") + "?" + c.Request.URL.RawQuery
	} else {
		url = "https://chat.openai.com/backend-api" + c.Param("path")
	}
	requestMethod = c.Request.Method

	request, err = http.NewRequest(requestMethod, url, c.Request.Body)
	if err != nil {
		log.Error("NewRequest Error: ", err.Error())
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}
	XAuthorization := c.Request.Header.Get("Authorization")

	request.Header.Set("Host", "chat.openai.com")
	request.Header.Set("Origin", "https://chat.openai.com/chat")
	request.Header.Set("Connection", "keep-alive")
	request.Header.Set("Content-Type", "application/json")
	request.Header.Set("Keep-Alive", "timeout=360")
	request.Header.Set("Authorization", XAuthorization)
	request.Header.Set("sec-ch-ua", "\"Chromium\";v=\"112\", \"Brave\";v=\"112\", \"Not:A-Brand\";v=\"99\"")
	request.Header.Set("sec-ch-ua-mobile", "?0")
	request.Header.Set("sec-ch-ua-platform", "\"Linux\"")
	request.Header.Set("sec-fetch-dest", "empty")
	request.Header.Set("sec-fetch-mode", "cors")
	request.Header.Set("sec-fetch-site", "same-origin")
	request.Header.Set("sec-gpc", "1")
	request.Header.Set("user-agent", userAgent)
	if Puid != "" {
		request.Header.Set("cookie", "_puid="+Puid+";")
	}

	response, err = client.Do(request)
	if err != nil {
		log.Error("Request Error: ", err.Error())
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}
	defer response.Body.Close()
	c.Header("Content-Type", response.Header.Get("Content-Type"))
	// Get status code
	c.Status(response.StatusCode)

	buf := make([]byte, 4096)
	for {
		n, err := response.Body.Read(buf)
		if n > 0 {
			_, writeErr := c.Writer.Write(buf[:n])
			if writeErr != nil {
				log.Error("Error writing to client: %v", writeErr)
				break
			}
			c.Writer.Flush()
		}
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Error("Error reading from response body: %v", err)
			break
		}
	}
}


api/api_proxy/chatgpt_api_proxy.go
package api_proxy

import (
	"crypto/tls"
	"fmt"
	log "github.com/popcell/chatgpt-proxy/utils/logger"
	"golang.org/x/net/context"
	"golang.org/x/net/proxy"
	"net"
	"net/http"
	"net/http/httputil"
	"net/url"
	"strings"
	"time"
)

var (
	ReverseProxy   *httputil.ReverseProxy
	proxyTransport = &http.Transport{
		TLSClientConfig:       &tls.Config{InsecureSkipVerify: true},
		DisableKeepAlives:     true,
		ResponseHeaderTimeout: time.Second * 60,
		IdleConnTimeout:       time.Second * 60,
		TLSHandshakeTimeout:   time.Second * 60,
	}
)

func InitApiProxyConfig(proxyUrl string, apiPrefix string) {

	if strings.HasPrefix(proxyUrl, "http") {
		proxyTransport.Proxy =
			func(req *http.Request) (*url.URL, error) {
				return url.Parse(proxyUrl)
			}
	} else if strings.HasPrefix(proxyUrl, "socks5") {
		u, err := url.Parse(proxyUrl)
		if err != nil {
			log.Error("InitApiProxyConfig error: ", err.Error())
			return
		}
		username := u.User.Username()
		password, _ := u.User.Password()
		host := u.Hostname()
		address := fmt.Sprintf("%s:%s", host, u.Port())
		dialer, err := proxy.SOCKS5("tcp", address, &proxy.Auth{User: username, Password: password}, proxy.Direct)
		proxyTransport.DialContext = func(ctx context.Context, network, addr string) (net.Conn, error) {
			return dialer.Dial(network, addr)
		}
		if err != nil {
			log.Error("InitApiProxyConfig error: ", err.Error())
		}
	}
	target, err := url.Parse("https://api.openai.com")
	if err != nil {
		log.Error("ApiProxy error: ", err)
	}
	ReverseProxy = httputil.NewSingleHostReverseProxy(target)
	ReverseProxy.Transport = proxyTransport
	ReverseProxy.Director = func(req *http.Request) {
		req.Host = target.Host
		req.URL.Scheme = target.Scheme
		req.URL.Host = target.Host
		req.URL.Path = strings.Replace(req.URL.Path, apiPrefix, "", -1)
	}

	// 打印HTTP请求和响应的日志
	//ReverseProxy.ModifyResponse = func(resp *http.Response) error {
	//	// 打印HTTP响应的日志
	//	responseDump, err := httputil.DumpResponse(resp, true)
	//	if err != nil {
	//		log.Info("Failed to dump response: ", err)
	//	} else {
	//		log.Info("Response: ", string(responseDump))
	//	}
	//	return nil
	//}
}


main.go
package main

import (
	"flag"
	"fmt"
	"github.com/gin-gonic/gin"
	"github.com/popcell/chatgpt-proxy/api"
	"github.com/popcell/chatgpt-proxy/api/api_proxy"
	"github.com/popcell/chatgpt-proxy/api/openai_proxy"
	"github.com/popcell/chatgpt-proxy/middleware"
	log "github.com/popcell/chatgpt-proxy/utils/logger"
	"github.com/spf13/viper"
	"os"
)

var (
	isLogin string
)

func initFlag() {
	flag.StringVar(&isLogin, "c", "", "Use -c login to login")
	flag.StringVar(&isLogin, "cmd", "", "Use -c login to login")

	// 自定义帮助信息
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "Usage: %s [OPTIONS]\n\n", os.Args[0])
		fmt.Fprintln(os.Stderr, "Options:")
		flag.PrintDefaults()
	}

	// 解析命令行参数
	flag.Parse()
}
func main() {
	initFlag()
	// 设置配置文件名和路径
	viper.SetConfigName("config")
	viper.AddConfigPath(".")
	viper.SetConfigType("yaml")

	// 读取配置文件
	err := viper.ReadInConfig()
	if err != nil {
		log.Panic("Read config failed: ", err)
	}
	host := viper.GetString("host")
	port := viper.GetInt("port")

	proxyUrl := viper.GetString("proxy-url")

	openAiProxyEnable := viper.GetBool("openai-proxy.enable")
	openAiEmail := viper.GetString("openai-proxy.email")
	openAiPassword := viper.GetString("openai-proxy.password")
	puid := viper.GetString("openai-proxy.puid")
	openAiProxyPrefix := viper.GetString("openai-proxy.prefix")

	apiProxyEnable := viper.GetBool("api-proxy.enable")
	apiProxyPrefix := viper.GetString("api-proxy.prefix")

	// 输出解析结果
	if isLogin == "login" {
		api.InitConfig(proxyUrl)
		api.Login(openAiEmail, openAiPassword)
		return
	}

	gin.SetMode(gin.ReleaseMode)
	gin.ForceConsoleColor()
	router := gin.Default()
	router.Use(middleware.CORSMiddleware())
	router.GET("/", func(c *gin.Context) {
		c.String(200, "Hello,World!")
	})
	if openAiProxyEnable {
		openai_proxy.InitOpenAiConfig(proxyUrl, puid)
		router.Any(fmt.Sprintf("/%s/*path", openAiProxyPrefix), openai_proxy.OpenAiProxy)
		fmt.Printf("openAiProxy run on %s:%d/%s\n", host, port, openAiProxyPrefix)
	}
	if apiProxyEnable {
		api_proxy.InitApiProxyConfig(proxyUrl, apiProxyPrefix)
		var chatGptUrl = fmt.Sprintf("/%s/*path", apiProxyPrefix)
		router.Any(chatGptUrl, func(c *gin.Context) {
			//requestDump, err := httputil.DumpRequest(c.Request, true)
			//if err != nil {
			//	log.Info("Failed to dump request: ", err)
			//} else {
			//	log.Info("Request: ", time.Now().Format("2006-01-02 15:04:05"), string(requestDump))
			//}
			api_proxy.ReverseProxy.ServeHTTP(c.Writer, c.Request)
		})
		fmt.Printf("chatGptApiProxy run on %s:%d/%s\n", host, port, apiProxyPrefix)
	}

	err = router.Run(fmt.Sprintf("%s:%d", host, port))
	if err != nil {
		log.Error("run server err: ", err.Error())
	}
}
